Benchmarking

jeremylt opened the pull request:
This WIP PR converts the scripts from the CEED/Benchmarks repo to the libCEED repo in order to add benchmarking to libCEED.

jeremylt commented:
ToDo:
- Loop over different backends, change output folder based on backend
- Integrate with top level makefile

jeremylt commented:
@v-dobrev I wanted to put this on your radar. I am modifying your scripts to make this a tool internal to the library, but I would love to hear any thought you have on my modifications. Currently you can get a basic run with 'make benchmarks' but I need to better handle options, machine configs, and scripts to run against.

jeremylt commented:
I'm running some final checks, but this should be good to go.

v-dobrev commented:
@jeremylt, I added a short description about running `make benchmarks` from the top level, however the postprocessing part is broken at the moment.

I think some of the output of running `benchmarks.sh` has to go to the output files, so the postprocessing script can read that information, e.g. which backend this data is from.

Also, I think the target `all` should be changed to build everything - the library, the tests, and the examples.

I'll work on fixing the postprocessing next.

jeremylt commented:
@v-dobrev, I'll be working on this tomorrow so we can put up some performance plots for SIAM CSE.

jeremylt commented:
@jedbrown, @v-dobrev, I fixed the postprocessing. I also converted the Python 2 scripts to Python 3 with EOL coming for Python 2. I think this PR is ready.

YohannDudouit commented:
It would be nice in the near future that the PETSc examples avoid moving data back and forth for GPU backends, I believe that it was mentioned in a CEED telecon that it wouldn't require much effort.

jeremylt commented:
Agreed. I would like to work with someone on that for a subsuquent PR.

v-dobrev commented:
I tried the latest commit and it seems to work fine. Also, the python scripts seemed to work fine using python 2.7.10, so v3 does not seem to be required.

One concern is that `make benchmarks` takes some time to run - maybe we can cap the number of iterations in CG when running the benchmarks? Not sure how much that will speed things up.

jeremylt commented:
I set 'ksp_max_it' to 20. It helps, but I think that we can only do so much with the fact that benchmarking all compiled backends just takes time.

jedbrown commented on examples/petsc/Makefile:
```diff
@@ -1,4 +1,5 @@
 PETSc.pc := $(PETSC_DIR)/$(PETSC_ARCH)/lib/pkgconfig/PETSc.pc
+PETSc.pc := $(or $(wildcard $(PETSc.pc)),$(PETSC_DIR)/lib/pkgconfig/PETSc.pc)
```
In what circumstance is there a `PETSc.pc` in this directory? If we really need to add this, we can delete the line above.

v-dobrev replied:
The location `$(PETSC_DIR)/lib/pkgconfig/PETSc.pc` is what I need when using PETSc installed through spack, for example.

jedbrown replied:
Yes, but the change is to look for `PETSc.pc` in the current directory. Why that change?

jedbrown replied:
Whoops, my mistake. So the environment you're thinking of is that `PETSC_ARCH` is set for a prefix install where it should not be?

v-dobrev replied:
The intent is to support installations in either `$(PETSC_DIR)/$(PETSC_ARCH)` or in `$(PETSC_DIR)`.

I guess the line I added to support the latter is unnecessary if `PETSC_ARCH` is undefined or empty - the path will have a double slash but that should still work. I probably looked-up where `PETSc.pc` is (in a spack install) without looking at the first case and never realized that leaving `PETSC_ARCH` empty is an option.

I'll remove the added line - I'll just add a comment that `PETSC_ARCH` can be empty.

jedbrown replied:
Perfect, thanks!

jedbrown commented on Makefile:
```diff
@@ -45,6 +45,11 @@ MAGMA_DIR ?= ../magma
                $(shell which nvcc 2> /dev/null))))),/usr/local/cuda)
 
+# Check for PETSc in ../petsc
+ifneq ($(wildcard ../petsc/lib/libpetsc.*),)
```
Is it ever installed like this?

v-dobrev replied:
This works when `../petsc` is (a link to) the directory where PETSc was installed (at least when installed through spack).

Clearly, this logic can be improved. What is a good way to check for a local build in that location? For example, what's a good way to get `PETSC_ARCH` automatically from such local build location?

v-dobrev replied:
@jedbrown, using the check as it is was convenient for me and it is consistent with how mfem, libxsmm, occa, and magma are setup up be auto-detected in `../<package-name>`. If you do not like it, feel free to modify it any way you like.

jedbrown replied:
But those packages have default builds (just run `make`) that leave the libraries and objects sitting next to the source. PETSc doesn't work that way so the user has to do something unnatural (create a special symlink; not a recommended method by PETSc) to use that code. Why didn't the old logic work for you?

v-dobrev replied:
This section is all new - there was no default setting for `PETSC_DIR` before, so I always had to add it to the make command line or through the environment. Both of these approaches still work. The addition is to add auto-detection for PETSc for my setup which seems reasonable for other people to use too. However, as I already said, feel free to remove it if you do not like it, especially if you think it may break something.

jedbrown replied:
Would your environment find `pkg-config PETSc`? (I.e., installed to a system path or anything in `PKG_CONFIG_PATH`?) How does any other PETSc-using package in your environment find it? I want to be sure there is no ambiguity to the user about which PETSc will be found and also that users don't confuse build and install directories in trying to conform to a directory layout that may only exist on your machine.

v-dobrev replied:
I do not setup `PKG_CONFIG_PATH` so `pkg-config PETSc` won't work for me right now.

When I need to use PETSc in MFEM, I set `PETSC_DIR` in my MFEM configuration file.

jedbrown commented on Makefile:
```diff
@@ -170,7 +178,7 @@ quiet = $(if $(V),$($(1)),$(call output,$1,$@);$($(1)))
 lib: $(libceed) $(ceed.pc)
 # run 'lib' target in parallel
-all:;@$(MAKE) $(MFLAGS) V=$(V) lib
+par:;@$(MAKE) $(MFLAGS) V=$(V) lib
```
Now phony target? Can we delete or make into a meaningful name like `lib-parallel`?

v-dobrev replied:
I wanted to use the target `all` for building the library with all the examples and all the tests - I think that is a more appropriate use of that target.

That is why, I just renamed the target to something short `all` -> `par` -- I think @camierjs was the main user of this target and I assumed he'd prefer something short to type. I guess I forgot to make it a phony target -- I'll do that in a moment.

jedbrown replied:
I'm not a fan of Makefile cleverness/abbreviations, but if someone is still attached to this, I won't argue.

jedbrown commented on Makefile:
```diff
@@ -93,6 +98,9 @@ NPROCS := $(shell getconf _NPROCESSORS_ONLN)
                        --no-print-directory --no-keep-going
 
+# Benchmark options
+BENCHMARK_EX := petsc-bp1
```
What about handling this more like tests, where there are `run-*` targets for each test?

jeremylt replied:
I'm investigating. I didn't introduce the other three changes you commented on, so I'll leave that to @v-dobrev to reply.

jedbrown replied:
Yes, exactly.

v-dobrev replied:
@jeremylt, if you are also OK with the above proposal, I can add to the makefile.

What about the `benchmarks` target? What is everyone's preference: should we keep it shorter and make it depend just on `bench-petsc-bp1` or should we be more comprehensive and have it run `$(allbechmarks)`?

v-dobrev replied:
I just made the above changes and updated README.md.

jeremylt commented:
Thank you @v-dobrev
